cmd/gcp-controller-manager/dpwi/serviceaccounts/eventhandler.go

```diff
@@ -0,0 +1,194 @@
+	verifier               *Verifier
+	notifyConfigmapHandler func()
+	notifyNodeSyncer       func(obj interface{})
+}
+
+// NewEventHandler creates a new eventHandler.
+func NewEventHandler(
+	saInformer coreinformers.ServiceAccountInformer,
+	podInformer coreinformers.PodInformer,
+	verifier *Verifier,
+	notifyConfigmapHandler func(),
+	notifyNodeSyncer func(obj interface{}),
+) *eventHandler {
+	podIndexer := podInformer.Informer().GetIndexer()
+	podIndexer.AddIndexers(map[string]cache.IndexFunc{
+		indexByKSA: func(obj interface{}) ([]string, error) {
+			pod, ok := obj.(*core.Pod)
+			if !ok {
+				return nil, fmt.Errorf("invalid pod object from key %q", obj)
+			}
+			ksa := fmt.Sprintf("%s/%s", pod.ObjectMeta.Namespace, pod.Spec.ServiceAccountName)
+			return []string{ksa}, nil
+		},
+	})
+	eh := &eventHandler{
+		saIndexer:  saInformer.Informer().GetIndexer(),
+		podIndexer: podIndexer,
+		saQueue: workqueue.NewNamedRateLimitingQueue(workqueue.NewMaxOfRateLimiter(
+			workqueue.NewItemExponentialFailureRateLimiter(200*time.Millisecond, 1000*time.Second),
+		), saVerifierSAQueueName),
+		verifier:               verifier,
+		notifyConfigmapHandler: notifyConfigmapHandler,
+		notifyNodeSyncer:       notifyNodeSyncer,
+	}
+	saInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc:    eh.onSAAdd,
+		UpdateFunc: eh.onSAUpdate,
+		DeleteFunc: eh.onSADelete,
+	}, serviceAccountResyncPeriod)
+	return eh
+}
+
+func (eh *eventHandler) onSAAdd(obj interface{}) {
+	eh.enqueueSA(obj)
+}
+
+func (eh *eventHandler) onSAUpdate(obj, oldObj interface{}) {
+	eh.enqueueSA(obj)
+}
+
+func (eh *eventHandler) onSADelete(obj interface{}) {
+	eh.enqueueSA(obj)
+}
+
+func (eh *eventHandler) enqueueSA(obj interface{}) {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		klog.Errorf("internal error. Couldn't get key for ServiceAccount %+v: %v", obj, err)
+		return
+	}
+	eh.saQueue.AddRateLimited(key)
+}
+
+// Run starts saWorkers number of ServiceAccount workers to process KSA events.
+// The workers will continue to run and process items from the workqueue until receiving stopCh.
+func (eh *eventHandler) Run(saWorkers int, stopCh <-chan struct{}) {
+	for i := 0; i < saWorkers; i++ {
+		go wait.Until(eh.workSAQueue, time.Second, stopCh)
+	}
+}
+
+func (eh *eventHandler) workSAQueue() {
+	for eh.processNextSA() {
+	}
+}
+
+func (eh *eventHandler) processNextSA() bool {
+	key, quit := eh.saQueue.Get()
+	if quit {
+		return false
+	}
+	defer eh.saQueue.Done(key)
+
+	if err := eh.process(key.(string)); err != nil {
+		klog.Warningf("Re-queuing SA %q due to %v", key, err)
+		eh.saQueue.AddRateLimited(key)
+		return true
+	}
+	eh.saQueue.Forget(key)
+	return true
+}
+
+func (eh *eventHandler) process(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	ksa := ServiceAccount{namespace, name}
+	if !eh.verifier.isProcessed(ksa) {
```

Comment:
It seems like just calling VerifiedGSA already does this?

Reply:
I have a similar comment but I think to start with "Force" or not should be a decision hidden the verifier.  Verifier can decide to force verify if the `ksa` is not previously seen or the target GSA has been changed.  As discussed I continue to be nervous about HMS being called expectedly by verifier users as `ForceVerify` is exposed, while `VerifiedGSA` also calls `ForceVerify` indirectly.  The decision to not call HMS if `isProcessed` and collapsing of calls to HMS with the same `ksa` by use of `DoChan` should always be there before inflicting any of the expensive calls to HMS.

Reply:
Deleted. Move the check inside Verifier.